### ISSUE TITLE
Allow ConditionalModule to execute `register` after evaluating its condition.

### ISSUE THREAD

**Issue by smehraein** [nestjs/nest#13710]
### Is there an existing issue that is already proposing this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe it

My application has a few different run configurations. Different env vars are required for different configurations. I attempt to load my modules using `ConditionalModule.registerWhen(Module1.register({options}), 'MODE_1')`, for example. However, inside `Module1.register`, I'm using `configService.getOrThrow`. Because the `register` is called immediately (not when the condition is evaluated), the application fails to start even when MODE_1 is false and I don't need that module or its configuration.

### Describe the solution you'd like

Allow `ConditionalModule.registerWhen` to accept a function which returns a module.

`ConditionalModule.registerWhen(() => Module1.register({options}), 'MODE_1)`

### Teachability, documentation, adoption, migration strategy

The documentation would be trivial. Just mention this option in the existing section on `ConditionalModule`.

### What is the motivation / use case for changing the behavior?

I want strict type-checking on my configuration, but I only want to perform those checks when the relevant modules are loaded.

**Comment by smehraein** [comment]
Perhaps this should be attached to the config module's repo? Apologies if so.

In any case, the code would be adjusted like this (based on this file: https://github.com/nestjs/config/blob/master/lib/conditional.module.ts).

First, the signature would be changed:
```
type CondModuleMetadata = Required<ModuleMetadata>['imports'][number];

static async registerWhen(
    module: CondModuleMetadata | () => CondModuleMetadata,
    condition: string | ((env: NodeJS.ProcessEnv) => boolean),
    options?: { timeout?: number; debug?: boolean },
  )
```

Then, on line 37, if the evaluation succeeds, check if `module` is a function, and evaluate it.

```
if (evaluation) {
     const returnModule = typeof module === 'function' ? module() : module;
      returnModule.imports.push(returnModule);
      returnModule.exports.push(returnModule);
    }
```

**Comment by micalevisk** [comment]
so we'd say that for such usage the module will be evaluated after `ConfigModule` succeeds on reading the env variables? This sounds reasonable to me

**Comment by smehraein** [comment]
That would be fantastic. I've experimented a bit and it seems like the solution is a bit more involved than I first guessed (even with the code above, it still seems to be evaluating the module).

And I have a couple of examples:

```
@Module({
  imports: [HttpModule.registerAsync({
     imports: [ConfigModule]
     inject: [ConfigService],
     useFactory: (config: ConfigService) => {baseUrl: config.getOrThrow(SPECIAL_URL)}
  })]
})
export class SubModuleA
```

Alternatively, with partial registration and config type-checking.

```
@Module({
  imports: [HttpModule.registerAsync({
     imports: [ConfigModule.forFeature(subModuleBConfig)]
     inject: [subModuleBConfig.KEY],
     useFactory: (config: ConfigType<typeof subModuleBConfig>) => {baseUrl: config.url}
  })]
})
export class SubModuleB
```
In this case, `subModuleBConfig` uses Zod to type-check the incoming config, and if SPECIAL_URL isn't there, will throw.

The important thing is that I'm deploying the same code in K8s with different environment variables to specify a run configuration. Depending on that top-level run-config variable, I'll need additional configurations. For example, if `ADMIN_MODE` is `true`, then I'd need `SPECIAL_URL`. If not, `SPECIAL_URL` can be safely omitted. One can imagine more complex cases where a module connects to some remote store for data, which is gated on the deployable's service account. I wouldn't want to execute that code unless I was running on the correct instance.

**Comment by micalevisk** [comment]
for those static ones I didn't understand why the current behavior of `ConditionalModule` isn't enough for your use case :thinking: I'd to see how exactly you're using it

**Comment by smehraein** [comment]
Sure, here's a specific example:

Inside the AppModule imports:
```
    ConditionalModule.registerWhen(CronModule, conditionMode(Server.CRON_MODE)),
```

```
// Allows us to default to FALSE when loading mode-specific modules.
function conditionMode(envVar: string) {
  return (env: NodeJS.ProcessEnv) => env[envVar]?.toLowerCase() === 'true';
}
```

`Server.CRON_MODE` is just `'CRON_MODE'` - I'm just namespacing my top-level env variable strings as constants.

Inside the CronModule (this normally has a separate file for the config, but it's quite simple here):

```
enum CronAction {
  REFRESH_LIST = 'REFRESH_LIST',
}

@Module({
  imports: [
    ScheduleModule.forRoot(),
    ConditionalModule.registerWhen(
      RefreshListsCronModule,
      conditionCronAction(CronAction.REFRESH_LIST),
    ),
  ],
})
export class CronModule {}

const cronConfigSchema = z.object({
  CRON_ACTION: z.nativeEnum(CronAction),
});

function conditionCronAction(action: CronAction) {
  return (env: NodeJS.ProcessEnv) => {
    const parsed = cronConfigSchema.parse(env);
    return action === parsed.CRON_ACTION;
  };
}
```

The objective is to run 1 instance for each family of crons (there's only one here, REFRESH_LISTS). That is governed by CRON_MODE = true and CRON_ACTION = 'REFRESH_LISTS'.

What happens here is that, even if CRON_MODE is false/unset (e.g. I'm running API_MODE), that `conditionCronAction` function will run and throw. The only way around it is to check *again* for CRON_MODE and safely exit the function. But that sort of defeats the purpose of ConditionalModule - I'd like to disable that entire section of code from AppModule.

**Comment by micalevisk** [comment]
ok, I got it now. You're not struggling with the module being evaluate but the condition, and you could end up parsing+validating the env obj multiple times (at `conditionCronAction` and by `ConfigModule` itself) just because `ConditionalModule.registerWhen` don't give you a chance to access `ConfigService` directly -- and this is by design.

So to me your use case seems a bit off from the feature set of `@nestjs/config`. Looks like you're trying to use it when it was not supposed to be used in first place.

As the proposed solution doesn't fix your use case, I'd rather not to add such feature. Instead we can figure out a better way to address your case at our [Discord server](http://discord.gg/nestjs)

**Comment by micalevisk** [comment]
if you want to keep using it, you'd have to handle the error raised by `cronConfigSchema.parse(env);`

I guess you can just ignore it if you're assuming that `@nestjs/config` will validate that later

**Comment by smehraein** [comment]
Actually, I think you're misunderstanding. The issue is that the module is being evaluated *at all*.

In the above example, my Cron module (and its children modules) require certain configuration values specific to their use case. The application should fail to start if those configs are invalid or missing. That list of required configs can be arbitrarily large and nested within the children of Cron module. That `conditionCronAction` function should *never* run if the CronModule isn't loaded.

However, if the Cron module isn't loaded, those configs are totally unnecessary.

The issue is that, because Cron module is being evaluated (even though it's not actually used, due to the ConditionalModule evaluating to false), I still need to have valid configurations for that entire subtree, even though I'm not using any of it. By extension, I need to provide valid configurations for every possible module and submodule for every configuration of the application.

This extends beyond just configuration - you could imagine establishing a database connection, or accessing a protected resource, which one version of the app doesn't have access to. That app should be able to disable the *entire* module sub-tree via Conditional module.

**Comment by smehraein** [comment]
Consider this example:
- Global Cron Module (Activated by: CRON_MODE=true) (Requires: CRON_ACTION)
  - CronSubmodule1 (Activated by: CRON_ACTION=1) (Requires: FOO_PARAM)
  - CronSubmodule1 (Activated by: CRON_ACTION=2) (Requires: BAR_PARAM)

If CRON_MODE is not true (unset), then none of CRON_ACTION, FOO_PARAM, or BAR_PARAM are required. We should not check or evaluate them.

If CRON_MODE is true, then CRON_ACTION must be set to '1' or '2'. Depending on the value of CRON_ACTION, either FOO_PARAM or BAR_PARAM is also required.

Because these checks are happening during module evaluation, it is currently required to set all 4 of these parameters to valid values to start the application, even if CRON_MODE is unset.

**Comment by smehraein** [comment]
Actually, in order to achieve the desired effect, CronModule *must* use a `register` function and return a dynamic module.

```
@Module({
  imports: [
   ...(process.env.CRON_MODE === 'true' ? [CronModule.register()] : [])
  ]
})
```

Anything within the `@Module` annotation seems to be evaluated regardless.

**Comment by micalevisk** [comment]
> Anything within the `@Module()` annotation seems to be evaluated regardless.

that's how TS decorators works. Not related with Nestjs at all.

Which is why the following statement didn't make sense and what I said before stands still

> ... That `conditionCronAction` function should never run if the CronModule isn't loaded.

---

Closing this as it seems that what you're requesting can be achieve either via the lazy-loading modules feature or by what you've mentioned already.

**Comment by smehraein** [comment]
For what it's worth, I was able to implement a custom version of this which fulfilled my needs. Up to you if you think it should be incorporated:

```
import { DynamicModule, ModuleMetadata } from '@nestjs/common';
import * as process from 'process';
import { ConditionalModule, ConfigModule } from '@nestjs/config';

type LazyModule = () => DynamicModule | Promise<DynamicModule>;

export class AsyncConditionalModule {
  /**
   * Registers the module if the provided function, evaluated against the environment, returns true.
   */
  static async registerWhen(
    module: LazyModule,
    envFn: (env: NodeJS.ProcessEnv) => boolean,
  ): Promise<DynamicModule> {
    await ConfigModule.envVariablesLoaded;
    const evaluation = envFn(process.env);
    return AsyncConditionalModule.register(module, evaluation);
  }

  private static register(
    module: LazyModule,
    condition: boolean,
  ): DynamicModule {
    if (condition) {
      const moduleActual = module();
      return {
        module: AsyncConditionalModule,
        imports: [moduleActual],
        exports: [moduleActual],
      };
    }
    return {
      module: AsyncConditionalModule,
    };
  }
}
```

(The separate, private `register` is there because I also added some (omitted) convenience functions for myself.

**Comment by micalevisk** [comment]
I think we can wait for more people to ask this as it is pretty easy to achieve and also leaves room for misunderstanding if one don't understand how nodejs modules works.

In your case you can write a `await import("module")` statement inside that function supplied for `registerWhen` to defer loading everything related with that module that is being loaded conditionally. So it's all about nodejs features, in the end.

Maybe we find a better/high level API to this limitation in the future.

Thanks!

**Comment by micalevisk** [comment]
Tbh I'm not a fan of `registerWhen` because it can make your app a bit hard to follow if you use this feature too much. I'd rather try to have different root modules that will be selected in the `boostrap` function depending on some condition.
